# Homework Assignment - Lesson 7: String Manipulation and Date/Time Data

**Student Name:** Carson Sloan

**Student ID:** 02149749

**Date Submitted:** 10/12/25

**Due Date:** 10/12/25

---

## Objective

Master string manipulation with `stringr` and date/time operations with `lubridate` for real-world business data cleaning and analysis.

## Learning Goals

By completing this assignment, you will:
- Clean and standardize messy text data using `stringr` functions
- Parse and manipulate dates using `lubridate` functions
- Extract information from text and dates for business insights
- Combine string and date operations for customer segmentation
- Create business-ready reports from raw data

## Instructions

- Complete all tasks in this notebook
- Write your code in the designated TODO sections
- Use the pipe operator (`%>%`) wherever possible
- Add comments explaining your logic
- Run all cells to verify your code works
- Answer all reflection questions

## Datasets

You will work with three CSV files:
- `customer_feedback.csv` - Customer reviews with messy text
- `transaction_log.csv` - Transaction records with dates
- `product_catalog.csv` - Product descriptions needing standardization

---

## Part 1: Data Import and Initial Exploration

**Business Context:** Before cleaning data, you must understand its structure and quality issues.

**Your Tasks:**
1. Load required packages (`tidyverse` and `lubridate`)
2. Import all three CSV files from the `data/` directory
3. Examine the structure and identify data quality issues
4. Display sample rows to understand the data

In [1]:
# Task 1.1: Load Required Packages
# TODO: Load tidyverse (includes stringr)
library(tidyverse)

# TODO: Load lubridate
library(lubridate)

cat("✅ Packages loaded successfully!\n")

── Attaching core tidyverse packages ──────────────────────── tidyverse 2.0.0 ──
✔ dplyr     1.1.4     ✔ readr     2.1.5
✔ forcats   1.0.0     ✔ stringr   1.5.2
✔ ggplot2   4.0.0     ✔ tibble    3.3.0
✔ lubridate 1.9.4     ✔ tidyr     1.3.1
✔ purrr     1.1.0     
── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()
ℹ Use the conflicted package (<http://conflicted.r-lib.org/>) to force all conflicts to become errors


✅ Packages loaded successfully!


In [3]:
# Task 1.2: Import Datasets

# Load tidyverse if not already loaded
library(tidyverse)

# TODO: Import customer_feedback.csv into a variable called 'feedback'
feedback <- read_csv("/workspaces/Carsons-stuff/data/customer_feedback.csv")

# TODO: Import transaction_log.csv into a variable called 'transactions'
transactions <- read_csv("/workspaces/Carsons-stuff/data/retail_transactions.csv")

# TODO: Import product_catalog.csv into a variable called 'products'
products <- read_csv("/workspaces/Carsons-stuff/data/product_catalog.csv")

# Print confirmation and row counts
cat("✅ Data imported successfully!\n")
cat("Feedback rows:", nrow(feedback), "\n")
cat("Transaction rows:", nrow(transactions), "\n")
cat("Product rows:", nrow(products), "\n")


Rows: 100 Columns: 5
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr  (2): Feedback_Text, Contact_Info
dbl  (2): FeedbackID, CustomerID
date (1): Feedback_Date

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
Rows: 500 Columns: 9


── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr  (4): CustomerName, CustomerCity, ProductName, ProductCategory
dbl  (4): TransactionID, CustomerID, TotalAmount, Quantity
date (1): TransactionDate

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.
Rows: 75 Columns: 5
── Column specification ────────────────────────────────────────────────────────
Delimiter: ","
chr (3): Product_Description, Category, In_Stock
dbl (2): ProductID, Price

ℹ Use `spec()` to retrieve the full column specification for this data.
ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.


✅ Data imported successfully!
Feedback rows: 100 
Transaction rows: 500 
Product rows: 75 


In [4]:
# Task 1.3: Initial Data Exploration

cat("=== CUSTOMER FEEDBACK DATA ===\n")
# TODO: Display structure of feedback using str()
str(feedback)

# TODO: Display first 5 rows of feedback
head(feedback, 5)

cat("\n=== TRANSACTION DATA ===\n")
# TODO: Display structure of transactions
str(transactions)

# TODO: Display first 5 rows of transactions
head(transactions, 5)

cat("\n=== PRODUCT CATALOG DATA ===\n")
# TODO: Display structure of products
str(products)

# TODO: Display first 5 rows of products
head(products, 5)



=== CUSTOMER FEEDBACK DATA ===
spc_tbl_ [100 × 5] (S3: spec_tbl_df/tbl_df/tbl/data.frame)
 $ FeedbackID   : num [1:100] 1 2 3 4 5 6 7 8 9 10 ...
 $ CustomerID   : num [1:100] 12 40 34 1 47 13 13 37 49 23 ...
 $ Feedback_Text: chr [1:100] "Highly recommend this item" "Excellent service" "Poor quality control" "average product, nothing special" ...
 $ Contact_Info : chr [1:100] "bob.wilson@test.org" "555-123-4567" "jane_smith@company.com" "jane_smith@company.com" ...
 $ Feedback_Date: Date[1:100], format: "2024-02-23" "2024-01-21" ...
 - attr(*, "spec")=
  .. cols(
  ..   FeedbackID = col_double(),
  ..   CustomerID = col_double(),
  ..   Feedback_Text = col_character(),
  ..   Contact_Info = col_character(),
  ..   Feedback_Date = col_date(format = "")
  .. )
 - attr(*, "problems")=<externalptr> 


FeedbackID CustomerID Feedback_Text                    Contact_Info          
1 1          12         Highly recommend this item       bob.wilson@test.org   
2 2          40         Excellent service                555-123-4567          
3 3          34         Poor quality control             jane_smith@company.com
4 4           1         average product, nothing special jane_smith@company.com
5 5          47         AMAZING customer support!!!      555-123-4567          
  Feedback_Date
1 2024-02-23   
2 2024-01-21   
3 2023-09-02   
4 2023-08-21   
5 2023-04-24


=== TRANSACTION DATA ===
spc_tbl_ [500 × 9] (S3: spec_tbl_df/tbl_df/tbl/data.frame)
 $ TransactionID  : num [1:500] 1 2 3 4 5 6 7 8 9 10 ...
 $ CustomerID     : num [1:500] 81 13 18 76 86 37 45 11 13 55 ...
 $ CustomerName   : chr [1:500] "Customer 39" "Customer 63" "Customer 98" "Customer 39" ...
 $ CustomerCity   : chr [1:500] "Chicago" "Philadelphia" "Chicago" "Houston" ...
 $ ProductName    : chr [1:500] "Adidas Jacket" "Samsung TV" "Adidas Jacket" "Dell Laptop" ...
 $ ProductCategory: chr [1:500] "Clothing" "Music" "Computers" "Computers" ...
 $ TotalAmount    : num [1:500] 632 114 1289 885 96 ...
 $ Quantity       : num [1:500] 3 3 7 2 5 2 3 3 8 1 ...
 $ TransactionDate: Date[1:500], format: "2024-03-09" "2024-12-08" ...
 - attr(*, "spec")=
  .. cols(
  ..   TransactionID = col_double(),
  ..   CustomerID = col_double(),
  ..   CustomerName = col_character(),
  ..   CustomerCity = col_character(),
  ..   ProductName = col_character(),
  ..   ProductCategory = col_character(),
  

TransactionID CustomerID CustomerName CustomerCity ProductName  
1 1             81         Customer 39  Chicago      Adidas Jacket
2 2             13         Customer 63  Philadelphia Samsung TV   
3 3             18         Customer 98  Chicago      Adidas Jacket
4 4             76         Customer 39  Houston      Dell Laptop  
5 5             86         Customer 45  New York     Nike Shoes   
  ProductCategory TotalAmount Quantity TransactionDate
1 Clothing         632.39     3        2024-03-09     
2 Music            114.28     3        2024-12-08     
3 Computers       1289.24     7        2024-01-22     
4 Computers        885.40     2        2024-07-02     
5 Computers         95.95     5        2024-08-13


=== PRODUCT CATALOG DATA ===
spc_tbl_ [75 × 5] (S3: spec_tbl_df/tbl_df/tbl/data.frame)
 $ ProductID          : num [1:75] 1 2 3 4 5 6 7 8 9 10 ...
 $ Product_Description: chr [1:75] "Apple iPhone 14 Pro - 128GB - Space Black" "samsung galaxy s23 ultra 256gb" "Apple iPhone 14 Pro - 128GB - Space Black" "Apple iPhone 14 Pro - 128GB - Space Black" ...
 $ Category           : chr [1:75] "TV" "TV" "Audio" "Shoes" ...
 $ Price              : num [1:75] 964 1817 853 649 586 ...
 $ In_Stock           : chr [1:75] "Limited" "Yes" "Yes" "Yes" ...
 - attr(*, "spec")=
  .. cols(
  ..   ProductID = col_double(),
  ..   Product_Description = col_character(),
  ..   Category = col_character(),
  ..   Price = col_double(),
  ..   In_Stock = col_character()
  .. )
 - attr(*, "problems")=<externalptr> 


ProductID Product_Description                       Category    Price  
1 1         Apple iPhone 14 Pro - 128GB - Space Black TV           963.53
2 2         samsung galaxy s23 ultra 256gb            TV          1817.44
3 3         Apple iPhone 14 Pro - 128GB - Space Black Audio        852.79
4 4         Apple iPhone 14 Pro - 128GB - Space Black Shoes        648.58
5 5         samsung galaxy s23 ultra 256gb            Electronics  586.35
  In_Stock
1 Limited 
2 Yes     
3 Yes     
4 Yes     
5 Limited

## Part 2: String Cleaning and Standardization

**Business Context:** Product names and feedback text often have inconsistent formatting that prevents accurate analysis.

**Your Tasks:**
1. Clean product names (remove extra spaces, standardize case)
2. Standardize product categories
3. Clean customer feedback text
4. Extract customer names from feedback

**Key Functions:** `str_trim()`, `str_squish()`, `str_to_lower()`, `str_to_upper()`, `str_to_title()`

In [7]:
# Task 2.1: Clean Product Names

library(stringr)  # for string manipulation functions

products_clean <- products %>%
  mutate(
    # Remove extra spaces and convert to Title Case
    product_name_clean = str_to_title(str_trim(Product_Description))
  )

# Display before and after
cat("Product Name Cleaning Results:\n")
products_clean %>%
  select(Product_Description, product_name_clean) %>%
  head(10) %>%
  print()


Product Name Cleaning Results:
# A tibble: 10 × 2
   Product_Description                         product_name_clean               
   <chr>                                       <chr>                            
 1 "Apple iPhone 14 Pro - 128GB - Space Black" "Apple Iphone 14 Pro - 128gb - S…
 2 "samsung galaxy s23 ultra 256gb"            "Samsung Galaxy S23 Ultra 256gb" 
 3 "Apple iPhone 14 Pro - 128GB - Space Black" "Apple Iphone 14 Pro - 128gb - S…
 4 "Apple iPhone 14 Pro - 128GB - Space Black" "Apple Iphone 14 Pro - 128gb - S…
 5 "samsung galaxy s23 ultra 256gb"            "Samsung Galaxy S23 Ultra 256gb" 
 6 "Apple iPhone 14 Pro - 128GB - Space Black" "Apple Iphone 14 Pro - 128gb - S…
 7 "DELL XPS 13 Laptop - Intel i7 - 16GB RAM"  "Dell Xps 13 Laptop - Intel I7 -…
 8 "hp envy printer - wireless - color"        "Hp Envy Printer - Wireless - Co…
 9 "Nike Air Max 270 - Size 10 - Black/White"  "Nike Air Max 270 - Size 10 - Bl…
10 "LG 55\" 4K Smart TV - OLED Display"        "Lg 55\" 4k 

In [8]:
# Task 2.2: Standardize Product Categories

library(stringr)  # for string manipulation

products_clean <- products_clean %>%
  mutate(
    # Remove extra spaces and convert to Title Case
    category_clean = str_to_title(str_trim(Category))
  )

# Show unique categories before and after
cat("Original categories:\n")
print(unique(products$Category))

cat("\nCleaned categories:\n")
print(unique(products_clean$category_clean))


Original categories:
[1] "TV"          "Audio"       "Shoes"       "Electronics" "Computers"  

Cleaned categories:
[1] "Tv"          "Audio"       "Shoes"       "Electronics" "Computers"  


In [12]:
# Task 2.3: Clean Customer Feedback Text

library(stringr)

feedback_clean <- feedback %>%
  mutate(
    # Convert text to lowercase and remove extra whitespace
    feedback_clean = str_squish(str_to_lower(Feedback_Text))
  )

# Display sample
cat("Feedback Cleaning Sample:\n")
feedback_clean %>%
  select(Feedback_Text, feedback_clean) %>%
  head(5) %>%
  print()



Feedback Cleaning Sample:
# A tibble: 5 × 2
  Feedback_Text                    feedback_clean                  
  <chr>                            <chr>                           
1 Highly recommend this item       highly recommend this item      
2 Excellent service                excellent service               
3 Poor quality control             poor quality control            
4 average product, nothing special average product, nothing special
5 AMAZING customer support!!!      amazing customer support!!!     


## Part 3: Pattern Detection and Extraction

**Business Context:** Identifying products with specific features and extracting specifications helps with inventory management and marketing.

**Your Tasks:**
1. Identify products with specific keywords (wireless, premium, gaming)
2. Extract numerical specifications from product names
3. Detect sentiment words in customer feedback
4. Extract email addresses from feedback

**Key Functions:** `str_detect()`, `str_extract()`, `str_count()`

In [13]:
# Task 3.1: Detect Product Features

library(stringr)

products_clean <- products_clean %>%
  mutate(
    # Create feature flags (case-insensitive detection)
    is_wireless = str_detect(str_to_lower(product_name_clean), "wireless"),
    is_premium  = str_detect(str_to_lower(product_name_clean), "pro|premium|deluxe"),
    is_gaming   = str_detect(str_to_lower(product_name_clean), "gaming|gamer")
  )

# Display results
cat("Product Feature Detection:\n")
products_clean %>%
  select(product_name_clean, is_wireless, is_premium, is_gaming) %>%
  head(10) %>%
  print()

# Summary statistics
cat("\nFeature Summary:\n")
cat("Wireless products:", sum(products_clean$is_wireless), "\n")
cat("Premium products:", sum(products_clean$is_premium), "\n")
cat("Gaming products:", sum(products_clean$is_gaming), "\n")


Product Feature Detection:
# A tibble: 10 × 4
   product_name_clean                          is_wireless is_premium is_gaming
   <chr>                                       <lgl>       <lgl>      <lgl>    
 1 "Apple Iphone 14 Pro - 128gb - Space Black" FALSE       TRUE       FALSE    
 2 "Samsung Galaxy S23 Ultra 256gb"            FALSE       FALSE      FALSE    
 3 "Apple Iphone 14 Pro - 128gb - Space Black" FALSE       TRUE       FALSE    
 4 "Apple Iphone 14 Pro - 128gb - Space Black" FALSE       TRUE       FALSE    
 5 "Samsung Galaxy S23 Ultra 256gb"            FALSE       FALSE      FALSE    
 6 "Apple Iphone 14 Pro - 128gb - Space Black" FALSE       TRUE       FALSE    
 7 "Dell Xps 13 Laptop - Intel I7 - 16gb Ram"  FALSE       FALSE      FALSE    
 8 "Hp Envy Printer - Wireless - Color"        TRUE        FALSE      FALSE    
 9 "Nike Air Max 270 - Size 10 - Black/White"  FALSE       FALSE      FALSE    
10 "Lg 55\" 4k Smart Tv - Oled Display"        FALSE       FALSE      FALS

In [16]:
# Task 3.2: Extract Product Specifications
# TODO: Create a new column 'size_number' that extracts the first number from product_name
# Hint: Use str_extract() with pattern "\\d+" to match one or more digits

products_clean <- products_clean %>%
  mutate(
    # Your code here:
    size_number = str_extract(Product_Description, "\\d+")
  )

# Display products with extracted sizes
cat("Extracted Product Specifications:\n")
products_clean %>%
  filter(!is.na(size_number)) %>%
  select(product_name_clean, size_number) %>%
  head(10) %>%
  print()

Extracted Product Specifications:
# A tibble: 10 × 2
   product_name_clean                          size_number
   <chr>                                       <chr>      
 1 "Apple Iphone 14 Pro - 128gb - Space Black" 14         
 2 "Samsung Galaxy S23 Ultra 256gb"            23         
 3 "Apple Iphone 14 Pro - 128gb - Space Black" 14         
 4 "Apple Iphone 14 Pro - 128gb - Space Black" 14         
 5 "Samsung Galaxy S23 Ultra 256gb"            23         
 6 "Apple Iphone 14 Pro - 128gb - Space Black" 14         
 7 "Dell Xps 13 Laptop - Intel I7 - 16gb Ram"  13         
 8 "Nike Air Max 270 - Size 10 - Black/White"  270        
 9 "Lg 55\" 4k Smart Tv - Oled Display"        55         
10 "Sony Wh-1000xm4 Wireless Headphones"       1000       


In [17]:
# Task 3.3: Simple Sentiment Analysis
# TODO: Create three new columns:
#   - positive_words: count of positive words ("great", "excellent", "love", "amazing")
#   - negative_words: count of negative words ("bad", "terrible", "hate", "awful")
#   - sentiment_score: positive_words - negative_words
# Hint: Use str_count() to count pattern occurrences

feedback_clean <- feedback_clean %>%
  mutate(
    # Your code here:
    positive_words = str_count(feedback_clean, "great|excellent|love|amazing"),
    negative_words = str_count(feedback_clean, "bad|terrible|hate|awful"),
    sentiment_score = positive_words - negative_words
  )

# Display sentiment analysis results
cat("Sentiment Analysis Results:\n")
feedback_clean %>%
  select(feedback_clean, positive_words, negative_words, sentiment_score) %>%
  head(10) %>%
  print()

# Summary
cat("\nOverall Sentiment Summary:\n")
cat("Average sentiment score:", mean(feedback_clean$sentiment_score), "\n")
cat("Positive reviews:", sum(feedback_clean$sentiment_score > 0), "\n")
cat("Negative reviews:", sum(feedback_clean$sentiment_score < 0), "\n")

Sentiment Analysis Results:
# A tibble: 10 × 4
   feedback_clean                  positive_words negative_words sentiment_score
   <chr>                                    <int>          <int>           <int>
 1 highly recommend this item                   0              0               0
 2 excellent service                            1              0               1
 3 poor quality control                         0              0               0
 4 average product, nothing speci…              0              0               0
 5 amazing customer support!!!                  1              0               1
 6 amazing customer support!!!                  1              0               1
 7 average product, nothing speci…              0              0               0
 8 good value for money                         0              0               0
 9 highly recommend this item                   0              0               0
10 highly recommend this item                   0             

## Part 4: Date Parsing and Component Extraction

**Business Context:** Transaction dates need to be parsed and analyzed to understand customer behavior patterns.

**Your Tasks:**
1. Parse transaction dates from text to Date objects
2. Extract date components (year, month, day, weekday)
3. Identify weekend vs weekday transactions
4. Extract quarter and month names

**Key Functions:** `ymd()`, `mdy()`, `dmy()`, `year()`, `month()`, `day()`, `wday()`, `quarter()`

In [21]:
# Task 4.1: Parse Transaction Dates
# TODO: Create a new column 'date_parsed' that parses the transaction_date column
# Hint: Check the format of transaction_date first, then use ymd(), mdy(), or dmy()

transactions_clean <- transactions %>%
  mutate(
    # Your code here:
    date_parsed = ymd(TransactionDate)
  )

# Verify parsing worked
cat("Date Parsing Results:\n")
transactions_clean %>%
  select(TransactionDate, date_parsed) %>%
  head(10) %>%
  print()

Date Parsing Results:
# A tibble: 10 × 2
   TransactionDate date_parsed
   <date>          <date>     
 1 2024-03-09      2024-03-09 
 2 2024-12-08      2024-12-08 
 3 2024-01-22      2024-01-22 
 4 2024-07-02      2024-07-02 
 5 2024-08-13      2024-08-13 
 6 2024-04-15      2024-04-15 
 7 2024-05-02      2024-05-02 
 8 2024-04-30      2024-04-30 
 9 2024-08-08      2024-08-08 
10 2024-06-23      2024-06-23 


In [22]:
# Task 4.2: Extract Date Components
# TODO: Create the following new columns:
#   - trans_year: Extract year from date_parsed
#   - trans_month: Extract month number from date_parsed
#   - trans_month_name: Extract month name (use label=TRUE, abbr=FALSE)
#   - trans_day: Extract day of month from date_parsed
#   - trans_weekday: Extract weekday name (use label=TRUE, abbr=FALSE)
#   - trans_quarter: Extract quarter from date_parsed

transactions_clean <- transactions_clean %>%
  mutate(
    # Your code here:
    trans_year = year(date_parsed),
    trans_month = month(date_parsed),
    trans_month_name = month(date_parsed, label = TRUE, abbr = FALSE),
    trans_day = day(date_parsed),
    trans_weekday = wday(date_parsed, label = TRUE, abbr = FALSE),
    trans_quarter = quarter(date_parsed)
  )

# Display results
cat("Date Component Extraction:\n")
transactions_clean %>%
  select(date_parsed, trans_month_name, trans_weekday, trans_quarter) %>%
  head(10) %>%
  print()

Date Component Extraction:
# A tibble: 10 × 4
   date_parsed trans_month_name trans_weekday trans_quarter
   <date>      <ord>            <ord>                 <int>
 1 2024-03-09  March            Saturday                  1
 2 2024-12-08  December         Sunday                    4
 3 2024-01-22  January          Monday                    1
 4 2024-07-02  July             Tuesday                   3
 5 2024-08-13  August           Tuesday                   3
 6 2024-04-15  April            Monday                    2
 7 2024-05-02  May              Thursday                  2
 8 2024-04-30  April            Tuesday                   2
 9 2024-08-08  August           Thursday                  3
10 2024-06-23  June             Sunday                    2


In [23]:
# Task 4.3: Identify Weekend Transactions
# TODO: Create a new column 'is_weekend' that is TRUE if the transaction was on Saturday or Sunday
# Hint: Use wday() which returns 1 for Sunday and 7 for Saturday
# Hint: Use %in% c(1, 7) to check if day is weekend

transactions_clean <- transactions_clean %>%
  mutate(
    # Your code here:
    is_weekend = wday(date_parsed) %in% c(1, 7)
  )

# Summary
cat("Weekend vs Weekday Transactions:\n")
table(transactions_clean$is_weekend) %>% print()

cat("\nPercentage of weekend transactions:",
    round(sum(transactions_clean$is_weekend) / nrow(transactions_clean) * 100, 1), "%\n")

Weekend vs Weekday Transactions:

FALSE  TRUE 
  363   137 

Percentage of weekend transactions: 27.4 %


## Part 5: Date Calculations and Customer Recency Analysis

**Business Context:** Understanding how recently customers transacted helps identify at-risk customers for re-engagement campaigns.

**Your Tasks:**
1. Calculate days since each transaction
2. Categorize customers by recency (Recent, Moderate, Old)
3. Identify customers who haven't transacted in 90+ days
4. Calculate average days between transactions per customer

**Key Functions:** `today()`, date arithmetic, `case_when()`

In [25]:
# Task 5.1: Calculate Days Since Transaction
# TODO: Create a new column 'days_since' that calculates days from date_parsed to today()
# Hint: Use as.numeric(today() - date_parsed)

transactions_clean <- transactions_clean %>%
  mutate(
    # Your code here:
    days_since = as.numeric(today() - date_parsed)
  )

# Display results
cat("Days Since Transaction:\n")
transactions_clean %>%
  select(CustomerName, date_parsed, days_since) %>%
  arrange(desc(days_since)) %>%
  head(10) %>%
  print()

Days Since Transaction:
# A tibble: 10 × 3
   CustomerName date_parsed days_since
   <chr>        <date>           <dbl>
 1 Customer 62  2024-01-01         650
 2 Customer 99  2024-01-01         650
 3 Customer 83  2024-01-02         649
 4 Customer 61  2024-01-02         649
 5 Customer 69  2024-01-02         649
 6 Customer 63  2024-01-03         648
 7 Customer 2   2024-01-03         648
 8 Customer 46  2024-01-03         648
 9 Customer 52  2024-01-05         646
10 Customer 15  2024-01-06         645


In [27]:
# Task 5.2: Categorize by Recency
# TODO: Create a new column 'recency_category' using case_when():
#   - "Recent" if days_since <= 30
#   - "Moderate" if days_since <= 90
#   - "At Risk" if days_since > 90

transactions_clean <- transactions_clean %>%
  mutate(
    # Your code here:
    recency_category = case_when(
      days_since <= 30 ~ "Recent",
      days_since <= 90 ~ "Moderate",
      days_since > 90 ~ "At Risk"
    )
  )

# Display distribution
cat("Recency Category Distribution:\n")
table(transactions_clean$recency_category) %>% print()

# Show at-risk customers
cat("\nAt-Risk Customers (>90 days):\n")
transactions_clean %>%
  filter(recency_category == "At Risk") %>%
  select(CustomerName, date_parsed, days_since) %>%
  arrange(desc(days_since)) %>%
  print()

Recency Category Distribution:

At Risk 
    500 

At-Risk Customers (>90 days):
# A tibble: 500 × 3
   CustomerName date_parsed days_since
   <chr>        <date>           <dbl>
 1 Customer 62  2024-01-01         650
 2 Customer 99  2024-01-01         650
 3 Customer 83  2024-01-02         649
 4 Customer 61  2024-01-02         649
 5 Customer 69  2024-01-02         649
 6 Customer 63  2024-01-03         648
 7 Customer 2   2024-01-03         648
 8 Customer 46  2024-01-03         648
 9 Customer 52  2024-01-05         646
10 Customer 15  2024-01-06         645
# ℹ 490 more rows


## Part 6: Combined String and Date Operations

**Business Context:** Create personalized customer outreach messages based on purchase recency.

**Your Tasks:**
1. Extract first names from customer names
2. Create personalized messages based on recency
3. Analyze transaction patterns by weekday
4. Identify best customers (recent + high value)

**Key Functions:** Combine `str_extract()`, date calculations, `case_when()`, `group_by()`, `summarize()`

In [28]:
# Task 6.1: Extract First Names and Create Personalized Messages
# TODO: Create two new columns:
#   - first_name: Extract first name from customer_name (everything before first space)
#   - personalized_message: Create message based on recency_category
#     * Recent: "Hi [name]! Thanks for your recent purchase!"
#     * Moderate: "Hi [name], we miss you! Check out our new products."
#     * At Risk: "Hi [name], it's been a while! Here's a special offer for you."
# Hint: Use str_extract() with pattern "^\\w+" for first name
# Hint: Use paste() to combine strings in case_when()

customer_outreach <- transactions_clean %>%
  mutate(
    # Your code here:
    first_name = str_extract(CustomerName, "^\\w+"),
    personalized_message = case_when(
      recency_category == "Recent" ~ paste("Hi", first_name, "! Thanks for your recent purchase!"),
      recency_category == "Moderate" ~ paste("Hi", first_name, ", we miss you! Check out our new products."),
      recency_category == "At Risk" ~ paste("Hi", first_name, ", it's been a while! Here's a special offer for you.")
    )
  )

# Display personalized messages
cat("Personalized Customer Messages:\n")
customer_outreach %>%
  select(CustomerName, first_name, days_since, personalized_message) %>%
  head(10) %>%
  print()

Personalized Customer Messages:
# A tibble: 10 × 4
   CustomerName first_name days_since personalized_message                      
   <chr>        <chr>           <dbl> <chr>                                     
 1 Customer 39  Customer          582 Hi Customer , it's been a while! Here's a…
 2 Customer 63  Customer          308 Hi Customer , it's been a while! Here's a…
 3 Customer 98  Customer          629 Hi Customer , it's been a while! Here's a…
 4 Customer 39  Customer          467 Hi Customer , it's been a while! Here's a…
 5 Customer 45  Customer          425 Hi Customer , it's been a while! Here's a…
 6 Customer 8   Customer          545 Hi Customer , it's been a while! Here's a…
 7 Customer 83  Customer          528 Hi Customer , it's been a while! Here's a…
 8 Customer 60  Customer          530 Hi Customer , it's been a while! Here's a…
 9 Customer 69  Customer          430 Hi Customer , it's been a while! Here's a…
10 Customer 24  Customer          476 Hi Customer , it's b

In [30]:
# Task 6.2: Analyze Transaction Patterns by Weekday
# TODO: Group by trans_weekday and calculate:
#   - transaction_count: number of transactions
#   - total_amount: sum of amount (if available)
#   - avg_amount: average amount per transaction
# TODO: Arrange by transaction_count descending

weekday_patterns <- transactions_clean %>%
  # Your code here:
  group_by(trans_weekday) %>%
  summarise(
    transaction_count = n()
  ) %>%
  arrange(desc(transaction_count))

# Display results
cat("Transaction Patterns by Weekday:\n")
print(weekday_patterns)

# Identify busiest day
busiest_day <- weekday_patterns$trans_weekday[1]
cat("\n🔥 Busiest day:", as.character(busiest_day), "\n")

Transaction Patterns by Weekday:
# A tibble: 7 × 2
  trans_weekday transaction_count
  <ord>                     <int>
1 Thursday                     81
2 Tuesday                      80
3 Saturday                     80
4 Wednesday                    75
5 Monday                       71
6 Sunday                       57
7 Friday                       56

🔥 Busiest day: Thursday 


In [31]:
# Task 6.3: Monthly Transaction Analysis
# TODO: Group by trans_month_name and calculate:
#   - transaction_count
#   - unique_customers: use n_distinct(customer_name)
# TODO: Arrange by trans_month (to show chronological order)

monthly_patterns <- transactions_clean %>%
  # Your code here:
  group_by(trans_month_name, trans_month) %>%
  summarise(
    transaction_count = n(),
    unique_customers = n_distinct(CustomerName),
    .groups = "drop"
  ) %>%
  arrange(trans_month)

# Display results
cat("Monthly Transaction Patterns:\n")
print(monthly_patterns)

Monthly Transaction Patterns:
# A tibble: 12 × 4
   trans_month_name trans_month transaction_count unique_customers
   <ord>                  <dbl>             <int>            <int>
 1 January                    1                42               39
 2 February                   2                43               36
 3 March                      3                41               34
 4 April                      4                43               36
 5 May                        5                32               28
 6 June                       6                40               32
 7 July                       7                47               35
 8 August                     8                43               34
 9 September                  9                35               30
10 October                   10                31               25
11 November                  11                43               37
12 December                  12                60               43


## Part 7: Business Intelligence Summary

**Business Context:** Create an executive summary that combines all your analyses into actionable insights.

**Your Tasks:**
1. Calculate key metrics across all datasets
2. Identify top products and categories
3. Summarize customer sentiment
4. Provide data-driven recommendations

In [32]:
# Task 7.1: Create Business Intelligence Dashboard

cat("\n", rep("=", 60), "\n")
cat("         BUSINESS INTELLIGENCE SUMMARY\n")
cat(rep("=", 60), "\n\n")

# Product Analysis
cat("📦 PRODUCT ANALYSIS\n")
cat(rep("─", 30), "\n")
# TODO: Calculate and display:
#   - Total number of products
#   - Number of wireless products
#   - Number of premium products
#   - Most common category

cat("Total Products:", nrow(products_clean), "\n")
cat("Wireless Products:", sum(products_clean$is_wireless, na.rm = TRUE), "\n")
cat("Premium Products:", sum(products_clean$is_premium, na.rm = TRUE), "\n")
most_common_cat <- names(sort(table(products_clean$category), decreasing = TRUE))[1]
cat("Most Common Category:", most_common_cat, "\n")


# Customer Sentiment
cat("\n💬 CUSTOMER SENTIMENT\n")
cat(rep("─", 30), "\n")
# TODO: Calculate and display:
#   - Total feedback entries
#   - Average sentiment score
#   - Percentage of positive reviews
#   - Percentage of negative reviews

cat("Total Feedback Entries:", nrow(feedback_clean), "\n")
cat("Average Sentiment Score:", round(mean(feedback_clean$sentiment_score), 2), "\n")
pct_positive <- round(sum(feedback_clean$sentiment_score > 0) / nrow(feedback_clean) * 100, 1)
pct_negative <- round(sum(feedback_clean$sentiment_score < 0) / nrow(feedback_clean) * 100, 1)
cat("Positive Reviews:", pct_positive, "%\n")
cat("Negative Reviews:", pct_negative, "%\n")


# Transaction Patterns
cat("\n📊 TRANSACTION PATTERNS\n")
cat(rep("─", 30), "\n")
# TODO: Calculate and display:
#   - Total transactions
#   - Date range (earliest to latest)
#   - Busiest weekday
#   - Weekend transaction percentage

cat("Total Transactions:", nrow(transactions_clean), "\n")
earliest_date <- min(transactions_clean$date_parsed, na.rm = TRUE)
latest_date <- max(transactions_clean$date_parsed, na.rm = TRUE)
cat("Date Range:", format(earliest_date, "%Y-%m-%d"), "to", format(latest_date, "%Y-%m-%d"), "\n")
busiest <- names(sort(table(transactions_clean$trans_weekday), decreasing = TRUE))[1]
cat("Busiest Weekday:", busiest, "\n")
weekend_pct <- round(sum(transactions_clean$is_weekend) / nrow(transactions_clean) * 100, 1)
cat("Weekend Transactions:", weekend_pct, "%\n")


# Customer Recency
cat("\n👥 CUSTOMER RECENCY\n")
cat(rep("─", 30), "\n")
# TODO: Calculate and display:
#   - Number of recent customers (< 30 days)
#   - Number of at-risk customers (> 90 days)
#   - Percentage needing re-engagement

recent_customers <- sum(transactions_clean$recency_category == "Recent", na.rm = TRUE)
at_risk_customers <- sum(transactions_clean$recency_category == "At Risk", na.rm = TRUE)
pct_at_risk <- round(at_risk_customers / nrow(transactions_clean) * 100, 1)

cat("Recent Customers (<30 days):", recent_customers, "\n")
cat("At-Risk Customers (>90 days):", at_risk_customers, "\n")
cat("Needing Re-engagement:", pct_at_risk, "%\n")



 = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = 
         BUSINESS INTELLIGENCE SUMMARY
= = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = = 

📦 PRODUCT ANALYSIS
─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ 
Total Products: 75 
Wireless Products: 17 
Premium Products: 13 


Warning message:
“Unknown or uninitialised column: `category`.”


Most Common Category: 

💬 CUSTOMER SENTIMENT
─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ 
Total Feedback Entries: 100 
Average Sentiment Score: 0.18 
Positive Reviews: 30 %
Negative Reviews: 20 %

📊 TRANSACTION PATTERNS
─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ 
Total Transactions: 500 
Date Range: 2024-01-01 to 2024-12-31 
Busiest Weekday: Thursday 
Weekend Transactions: 27.4 %

👥 CUSTOMER RECENCY
─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ ─ 
Recent Customers (<30 days): 0 
At-Risk Customers (>90 days): 500 
Needing Re-engagement: 100 %


In [33]:
# Task 7.2: Identify Top Products by Category
# TODO: Group products by category_clean and count products in each
# TODO: Arrange by count descending
# TODO: Display top 5 categories

top_categories <- products_clean %>%
  # Your code here:
  group_by(category_clean) %>%
  summarise(
    product_count = n()
  ) %>%
  arrange(desc(product_count)) %>%
  head(5)

cat("Top Product Categories:\n")
print(top_categories)

Top Product Categories:
# A tibble: 5 × 2
  category_clean product_count
  <chr>                  <int>
1 Electronics               21
2 Computers                 15
3 Audio                     14
4 Tv                        14
5 Shoes                     11


## Part 8: Reflection Questions

Answer the following questions based on your analysis. Write your answers in the markdown cells below.

### Question 8.1: Data Quality Impact

**How did cleaning the text data (removing spaces, standardizing case) improve your ability to analyze the data? Provide specific examples from your homework.**

Your answer here:
Cleaning text data enabled me to transform messy and inconsistent data into standardized formats for counting grouping, and filtering. When grouping by category_clean to find top product categories, we got accurate counts instead of splitting the same category across multiple rows


### Question 8.2: Pattern Detection Value

**What business insights did you gain from detecting patterns in product names (wireless, premium, gaming)? How could a business use this information?**

Your answer here:
Pattern discovery transforms unstructured product names to actionable business insights and enables data-driven decision making on product development, marketing, inventory, and competition. Businesses could use this to establish if premium products ought to have deserved price premiums. 


### Question 8.3: Date Analysis Importance

**Why is analyzing transaction dates by weekday and month important for business operations? Provide at least three specific business applications.**

Your answer here:

1. Businesses may need to analyze transactions by weekday and month for business operations for things like staff scheduling. Managers can use past information to avoid over-staffing on slow days and understand when to staff more. 
2. Businesses can also better manage inventory. Businesses will have more inventory and stock up during their busier seasons.
3. Businesses can also use this information to market more during their busier seasons or launch sales during their slower days. 


### Question 8.4: Customer Recency Strategy

**Based on your recency analysis, what specific actions would you recommend for customers in each category (Recent, Moderate, At Risk)? How would you prioritize these actions?**

Your answer here:



### Question 8.5: Sentiment Analysis Application

**How could the sentiment analysis you performed be used to improve products or customer service? What are the limitations of this simple sentiment analysis approach?**

Your answer here:
The sentiment analysis can be used to find products that arent doing well and have them pulled. This could also help find unhappy customers for damage control. A limitation of a simple sentiment analysis is that you cant really differentiate between mildly posiitve or negative reviews. 


### Question 8.6: Real-World Application

**Describe a real business scenario where you would need to combine string manipulation and date analysis (like you did in this homework). What insights would you be trying to discover?**

Your answer here:
Businesses would combine string manipulation and date analysis when looking at patterns between products and seasons. This would be like a surplus of gift sets around christmas or the increase in sales of video games leading into summer. Business would be trying to discover what products will sell most at what times so they can stock up for those seasons.


## Summary and Submission

### What You've Accomplished

In this homework, you've successfully:
- ✅ Cleaned and standardized messy text data using `stringr` functions
- ✅ Detected patterns and extracted information from text
- ✅ Parsed dates and extracted temporal components using `lubridate`
- ✅ Calculated customer recency for segmentation
- ✅ Analyzed transaction patterns by time periods
- ✅ Combined string and date operations for business insights
- ✅ Created personalized customer communications
- ✅ Generated executive-ready business intelligence summaries

### Key Skills Mastered

**String Manipulation:**
- `str_trim()`, `str_squish()` - Whitespace handling
- `str_to_lower()`, `str_to_upper()`, `str_to_title()` - Case conversion
- `str_detect()` - Pattern detection
- `str_extract()` - Information extraction
- `str_count()` - Pattern counting

**Date/Time Operations:**
- `ymd()`, `mdy()`, `dmy()` - Date parsing
- `year()`, `month()`, `day()`, `wday()` - Component extraction
- `quarter()` - Period extraction
- `today()` - Current date
- Date arithmetic - Calculating differences

**Business Applications:**
- Data cleaning and standardization
- Customer segmentation by recency
- Sentiment analysis
- Pattern identification
- Temporal trend analysis
- Personalized communication

### Submission Checklist

Before submitting, ensure you have:
- [ ] Entered your name, student ID, and date at the top
- [ ] Completed all code tasks (Parts 1-7)
- [ ] Run all cells successfully without errors
- [ ] Answered all reflection questions (Part 8)
- [ ] Used proper commenting in your code
- [ ] Used the pipe operator (`%>%`) where appropriate
- [ ] Verified your results make business sense
- [ ] Checked for any remaining TODO comments

### Grading Criteria

Your homework will be evaluated on:
- **Code Correctness (40%)**: All tasks completed correctly
- **Code Quality (20%)**: Clean, well-commented, efficient code
- **Business Understanding (20%)**: Demonstrates understanding of business applications
- **Reflection Questions (15%)**: Thoughtful, complete answers
- **Presentation (5%)**: Professional formatting and organization

### Next Steps

In Lesson 8, you'll learn:
- Advanced data wrangling with complex pipelines
- Sophisticated conditional logic with `case_when()`
- Data validation and quality checks
- Creating reproducible analysis workflows
- Professional best practices for business analytics

**Great work on completing this assignment! 🎉**